homeassistant/components/vicare/__init__.py

```diff
@@ -136,62 +135,52 @@ async def async_migrate_devices(
-        """Update unique ID of entity entry."""
-        if not entity_entry.unique_id.startswith(gateway_serial):
-            # belongs to other device/gateway
-            return None
-        if entity_entry.unique_id.startswith(f"{gateway_serial}_"):
-            # Already correct, nothing to do
-            return None
-
-        unique_id_parts = entity_entry.unique_id.split("-")
-        unique_id_parts[0] = new_identifier
-
-        # convert climate entity unique id from `<device_identifier>-<circuit_no>` to `<device_identifier>-heating-<circuit_no>`
-        if entity_entry.domain == DOMAIN_CLIMATE:
-            unique_id_parts[len(unique_id_parts) - 1] = (
-                f"{entity_entry.translation_key}-{unique_id_parts[len(unique_id_parts)-1]}"
-            )
-
-        entity_new_unique_id = "-".join(unique_id_parts)
-
-        _LOGGER.debug(
-            "Migrating entity %s from %s to new id %s",
-            entity_entry.entity_id,
-            entity_entry.unique_id,
-            entity_new_unique_id,
-        )
-        return {"new_unique_id": entity_new_unique_id}
-
-    # Migrate entities
-    await er.async_migrate_entries(hass, entry.entry_id, _update_unique_id)
+            @callback
+            def _update_unique_id(
+                entity_entry: er.RegistryEntry,
+            ) -> dict[str, str] | None:
+                """Update unique ID of entity entry."""
+                if entity_entry.device_id != device_entry.id:
+                    # belongs to another device
+                    return None
+                if not entity_entry.unique_id.startswith(gateway_serial):
+                    # belongs to other device/gateway
+                    return None
+                if entity_entry.unique_id.startswith(new_identifier):
+                    # already correct, nothing to do
+                    return None
+
+                unique_id_parts = entity_entry.unique_id.split("-")
+                # replace old prefix `<gateway-serial>` with `<gateways-serial>_<device-serial>`
+                unique_id_parts[0] = new_identifier
+                # convert climate entity unique id from `<device_identifier>-<circuit_no>` to `<device_identifier>-heating-<circuit_no>`
+                if entity_entry.domain == DOMAIN_CLIMATE:
+                    unique_id_parts[len(unique_id_parts) - 1] = (
+                        f"{entity_entry.translation_key}-{unique_id_parts[len(unique_id_parts)-1]}"
+                    )
+                entity_new_unique_id = "-".join(unique_id_parts)
+
+                _LOGGER.debug(
+                    "Migrating entity %s from %s to new id %s",
+                    entity_entry.entity_id,
+                    entity_entry.unique_id,
+                    entity_new_unique_id,
+                )
+                return {"new_unique_id": entity_new_unique_id}
+
+            # Migrate entities
+            await er.async_migrate_entries(hass, entry.entry_id, _update_unique_id)
```

Comment:
If we're looping over all the device entries of the config entry and want to migrate the corresponding entity entries, we shouldn't use this helper that migrates all the entity entries of the config entry.

Either look up the entity entries to migrate from each device entry or decouple the migration of the entity entries from the migration of the device entries.

Reply:
Changed to iterate over ViCare devices, then over HA devices and over entities related to that device. Can probably be improved by looking up the device directly. Will check if that's possible the next days..